File: 技术/R/dpmm_cgs/gmm_em.R
# 生产多维高斯混合分布数据
set.seed(410)

gene_data <- function(n,p){
  # n: length of data
  # p: dimention of x
  # rho: Covariance matrix correlation
  rho       <- runif(p,0,1)
  w         <- DIRECT::rDirichlet(1,rep(1,p)) # weights of every distribution
  x         <- data.frame()
  means     <- list()
  sigmas    <- list()
  for (i in (1:p)) {
    n_samples <- ceiling(n*w[i])
    mean      <- rnorm(p,0,1)
    sigma_mat <- function(p,rho){diag(1-rho,p,p)+rho}
    sigma     <- sigma_mat(p,rho[i])
    temp      <- MASS::mvrnorm(n_samples, mean, sigma)
    x         <- rbind(x,temp)
    means     <- c(means,list(mean))
    sigmas    <- c(sigmas,list(sigma))
  }
  return(list(x = x[1:n,],w = w,means = means,sigma = sigmas,rho = rho))
}

result <- gene_data(1000,5)

# EM算法
# https://www.jianshu.com/p/8acf6ec8193e

# 计算多元高斯分布概率值
Gaussian <- function(data,mean,cov){
  data = as.matrix(data)
  mean = as.numeric(mean)
  cov  = as.matrix(cov)
  
  dime   <- dim(cov)[1]
  covdet <- det(cov)
  covinv <- solve(cov)
  if (covdet == 0){
    covdet = det(cov+diag(dime)*0.01)
    covinv = solve(cov+diag(dime)*0.01)}
  m  = as.matrix(data - mean)
  z1 = -0.5*(m %*% covinv %*% t(m))
  z2 = ((((2*pi)^dime)*abs(covdet))^0.5)
  return(exp(z1)/z2)
}

gmm <- function(data,k){
  n     = dim(data)[1]
  # 设定初始迭代值（均值，协方差）
  convs = list()
  means = list()
  temp1 <- cov(data)
  temp2 <- sapply(data,mean)
  for (i in 1:k){
    convs[[i]] <- temp1 + rnorm(1,0,0.1)
    means[[i]] <- temp2 + rnorm(1,0,0.1)
  }
  # 初始权重
  pis    <- rep(1.0/k,k)
  gammas <- matrix(rep(0,k*n),n,k)
  
  loglikelihood    <- 0
  oldloglikelihood <- 1
  res   = matrix(rep(0,k*n),n,k)
  
  count = 0
  while (abs(loglikelihood-oldloglikelihood) > 0.0001 & count < 10){
    count = count +1
    cat("count = ",count,"  ","loglikelihood = ",loglikelihood,"\n")
    oldloglikelihood = loglikelihood
    # E step 求解gammas[n,k],即第n个样本落在第k个高斯分布的概率，对数似然
    for (i in 1:n) {
      for (j in 1:k){res[i,j] = pis[j]*Gaussian(data[i,],means[[j]],convs[[j]])}
      sumres <- sum(res[i,])
      for (j in 1:k){gammas[i,j] = res[i,j]/sumres}
    }
    # M step update parameters
    for (j in 1:k){
      Nj     <- sum(gammas[,j])
      pis[j] <- 1.0*Nj/n
      
      summea <- 0
      for (i in 1:n) {summea <- summea +  gammas[i,j]*data[i,]}
      means[j] <- list(summea/Nj)
      xdiffs   <- list()
      for (i in 1:n) {xdiffs[[i]] <- data[i,] - summea/Nj}
      sumcov   <- 0
      for (i in 1:n) {sumcov <- sumcov +  gammas[i,j]*t(as.matrix(xdiffs[[i]]))%*%as.matrix(xdiffs[[i]])}
      convs[j] <- list((1.0/Nj)*sumcov)
    }
    loglikelihood <- 0
    for (i in 1:n){
      temp <- 0
      for (j in 1:k){
        temp <- temp + pis[j]*Gaussian(data[i,],means[[j]],convs[[j]])
      }
      loglikelihood <- loglikelihood + log(temp)
    }
  } 
  return(list(loglikelihood = loglikelihood,pis = pis,means = means,convs = convs,gammas = gammas))
}

a = Sys.time()
resofgmm <- gmm(result$x,5)
print(Sys.time()-a)

# 数据分割及采样
mysample <- function(lst1,lst2){
  x = lst1$x
  k = dim(x)[2]
  gammas <- lst2$gammas
  newx = list()
  newgammas <- ismax(gammas)
  for (i in 1:k){
    xi = x[newgammas[,i]==1,]
    newx = c(newx,list(xi))
  }
  return(newx)
}

# 定义判断最大值矩阵
ismax <- function(d){
  l = dim(d)[1]
  p = dim(d)[2]
  for (i in 1:l) {
    ma = max(d[i,])
    for (j in 1:p) {d[i,j] <- if (d[i,j] == ma) TRUE else FALSE}
  }
  return(d)
}

res = mysample(result,resofgmm)

File: 技术/python/gmm/gmm_em.py
import time
a = time.time()

# 生成GMM数据
import numpy as np
def generate_data(n,p):
    '''
    计算服从均值为mean，协方差为cov的多元正态分布的数据
    :param mean: 均值
    :param cov: 协方差
    :return: x，服从GMM分布的数据
    '''
    global x
    global rho_list # 变量相关性
    global cov_list
    global mean_list
    global w
    
    w = np.random.random(p)
    w = w/sum(w)
    xx = list()
    rho_list = list()
    cov_list = list()
    mean_list = list()
    for i in range(p):
        #n_samples = int(n*w[i])
        n_samples = int(round(n*w[i]))
        mean = np.random.random(p)
        rho = np.random.random()
        cov = np.diag(np.repeat(1-rho,p))+rho
        temp = np.random.multivariate_normal(mean,cov,n_samples)
        rho_list.append(rho)
        cov_list.append(cov)
        mean_list.append(mean)
        xx.append(temp)
    x = np.vstack((xx[0],xx[1]))
    for i in range(2,p):
        x = np.vstack((x,xx[i]))
    return x
        
if __name__ == '__main__':
    p = 5    # 五维数据
    n = 500 # 数据长度
    generate_data(n,p) 
    
    


# 计算高斯函数
def Gaussian(data,mean,cov):
    dim = np.shape(cov)[0]      # 计算维度
    covdet = np.linalg.det(cov) # 计算|cov|
    covinv = np.linalg.inv(cov) # 计算cov的逆
    if covdet==0:               # 以防行列式为0
        covdet = np.linalg.det(cov+np.eye(dim)*0.01)
        covinv = np.linalg.inv(cov+np.eye(dim)*0.01)
    m = data - mean
    z = -0.5 * np.dot(np.dot(m, covinv),m)    # 计算exp()里的值
    # 返回概率密度值
    return 1.0/(np.power(np.power(2*np.pi,dim)*abs(covdet),0.5))*np.exp(z)


def GMM(data,K):
    N = data.shape[0]
    dim = data.shape[1]

    convs=[0]*dim
    means=[0]*dim

    # 初始方差等于整体data的方差
    for i in range(K):
        convs[i]=np.cov(data.T)+np.random.random()/10
        means[i]=np.mean(data.T)+np.random.random()/10
        #convs[i]=np.diag(np.repeat(1-0.1,p))+0.1
        #means[i]=np.random.random(dim)+np.mean(data.T)       
        # means[i]=np.zeros(dim)
    
    pis = [1.0/K] * K
    gammas = [np.zeros(K) for i in range(N)]
    
    loglikelyhood = 0
    oldloglikelyhood = 1

    while np.abs(loglikelyhood - oldloglikelyhood) > 0.0001:
        oldloglikelyhood = loglikelyhood
        
        # E步
        for i in range(N):
            res = [pis[k] * Gaussian(data[i],means[k],convs[k]) for k in range(K)]
            sumres = np.sum(res)
            for k in range(K):      # gamma表示第n个样本属于第k个混合高斯的概率
                gammas[i][k] = res[k] / sumres
        # M步
        for k in range(K):
            Nk = np.sum([gammas[n][k] for n in range(N)])  # N[k] 表示N个样本中有多少属于第k个高斯
            pis[k] = 1.0 * Nk/N
            means[k] = (1.0/Nk)*np.sum([gammas[n][k] * data[n] for n in range(N)],axis=0)
            xdiffs = data - means[k]
            convs[k] = (1.0/ Nk)*np.sum([gammas[n][k]* xdiffs[n].reshape(dim,1) * xdiffs[n] for  n in range(N)],axis=0)
        # 计算最大似然函数
        loglikelyhood = np.sum([np.log(np.sum([pis[k] * Gaussian(data[n], means[k], convs[k]) for k in range(K)])) for n in range(N)])
        print(loglikelyhood)
    print("pi :" ,end = " ")
    print(pis)
    print("means :")
    for item in means:
        print(item)
    print("convs :")
    for item in convs:
        print(item)
    print("loglikelyhood :",end = " ")
    print (loglikelyhood)
    
GMM(x,5)

b = time.time()
print(b-a)
